simple_certmanager/models.py

```diff
@@ -10,12 +10,53 @@
 )
 from privates.fields import PrivateMediaFileField
 
+from simple_certmanager.csr_generation import generate_csr_and_private_key
+
 from .constants import CertificateTypes
 from .mixins import DeleteFileFieldFilesMixin
 from .utils import load_pem_x509_private_key, pretty_print_certificate_components
 from .validators import PrivateKeyValidator, PublicCertValidator
 
 
+class SigningRequest(models.Model):
+    common_name = models.CharField(
+        max_length=100,
+        help_text=_("The common name for the certificate"),
+    )
+    country_name = models.CharField(
+        max_length=2, help_text=_("Two-letter country code"), blank=True, default="NL"
+    )
+    organization_name = models.CharField(
+        max_length=100, help_text=_("The name of the organization"), blank=True
+    )
+    state_or_province_name = models.CharField(
+        max_length=100, help_text=_("The state or province name"), blank=True
+    )
+    email_address = models.EmailField(
+        help_text=_("Email address for the certificate"), blank=True
+    )
+    csr = models.TextField(
+        editable=False,
+        blank=True,
+        help_text=_("Certificate Signing Request"),
+        verbose_name=_("CSR"),
+    )
+    private_key = models.TextField(
+        editable=False,
+        blank=True,
+    )
+
+    def save(self, *args, **kwargs):
+        generate_csr_and_private_key(self)
+        super().save(*args, **kwargs)
```

Comment:
with my proposed changes in `csr_generation.py`, this becomes:

```suggestion
    def save(self, *args, **kwargs):
        # we generate private key + csr as one atomic unit
        # TODO: add a check-constraint at the DB level to enforce this.
        if not self.private_key:
            self.private_key, self.csr = generate_private_key_with_csr(
                common_name=self.common_name,
                organization_name=self.organization_name,
                # XXX field for locality/city is missing!
                # locality=self.locality_name,
                state_or_province=self.state_or_province_name,
                country=self.country_name,
                email=self.email_address,
            )

        super().save(*args, **kwargs)
```